src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -48,28 +48,85 @@ public static void generateReposReport(List<RepoConfiguration> configs, String o
         InputStream is = RepoSense.class.getResourceAsStream(TEMPLATE_FILE);
         FileUtil.copyTemplate(is, outputPath);
 
-        for (RepoConfiguration config : configs) {
-            Path repoReportDirectory;
-            try {
-                GitClone.clone(config);
-                repoReportDirectory = Paths.get(outputPath, config.getDisplayName());
-                FileUtil.createDirectory(repoReportDirectory);
-            } catch (GitCloneException gde) {
-                logger.log(Level.WARNING,
-                        "Exception met while trying to clone the repo, will skip this repo.", gde);
-                repoReportDirectory = Paths.get(outputPath, config.getDisplayName());
-                FileUtil.createDirectory(repoReportDirectory);
-                generateEmptyRepoReport(repoReportDirectory.toString());
-                continue;
-            } catch (IOException ioe) {
-                logger.log(Level.WARNING,
-                        "Error has occurred while creating repo directory, will skip this repo.", ioe);
-                continue;
-            } catch (RuntimeException rte) {
-                logger.log(Level.SEVERE, "Error has occurred during analysis, will skip this repo.", rte);
-                continue;
+        cloneAndAnalyzeRepos(configs, outputPath);
+
+        try {
+            FileUtil.deleteDirectory(FileUtil.REPOS_ADDRESS);
+        } catch (IOException ioe) {
+            logger.log(Level.WARNING, "Error deleting report directory.", ioe);
+        }
+        FileUtil.writeJsonFile(new SummaryReportJson(configs, generationDate), getSummaryResultPath(outputPath));
+        logger.info("The report is generated at " + outputPath);
+    }
+
+    /**
+     * Clones and analyzes repositories in {@code configs}.
+     * Performs analysis of each repository in parallel with the cloning of the next repository.
+     */
+    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) throws IOException {
+        boolean isPreviousRepoCloned = false;
+
+        for (int i = 0; i < configs.size(); i++) {
+            RepoConfiguration config = configs.get(i);
+            boolean isCurrentRepoCloned = spawnCloneProcess(outputPath, config);
+
+            if (isPreviousRepoCloned) {
```

Comment:
Seems better, but we can improve it further. 
Let's have another class, say `RepoCloner` to take care of the cloning and cleaning up.

This `RepoCloner` can have hold up to 2 `RepoConfiguration` and always be responsibled for cleaning up the older one. 
So the instance variables will have something like this.
```
    RepoConfiguration[] configs = new RepoConfiguration[2];
    int index = 0;
    int prevIndex = 0;
``` 
So if we do something like `repoCloner.clone(config)` the backend should do something like this:
```
void clone(RepoConfiguration config) {
    prevIndex = index;
    if (config.getLocation().equals(configs[index])) {
        return;
    }
    index = index + 1 % configs.length;
    FileUtil.deleteDirectory(configs[index].getRepoRoot());
    spawnCloneProcess(outputPath, config);
}
```

We can then have `repoCloner` return the previous repo with `repoCloner.getClonedRepo()`
So backend would be something like this:
```
RepoConfiguration getClonedRepo() {
    waitForCloneProcess(configs[prevIndex]);
    Git.checkout(configs.getBranch()); 
    return configs[prevIndex];
}
```

Note that with this new change, we need to remember the default branch of the cloned repo. This is because it can result in error if the order of configs are something like this:|
```
reposense, release
reposense, empty -> default branch (HEAD)
```
This will result in the second configuration not analyzing the correct branch.